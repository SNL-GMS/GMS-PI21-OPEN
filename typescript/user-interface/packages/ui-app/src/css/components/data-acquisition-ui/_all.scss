@import 'soh-vars';
@import 'soh-cell';
@import 'ag-theme-overrides.scss';
@import 'data-acquisition';
@import 'environment-history';
@import 'historical-trends-history';
@import 'soh-chart';
@import 'station-statistics';
@import 'soh-lag-missing';
@import 'soh-environment';
@import 'soh-overview';
@import 'soh-quiet-indicator';
@import 'capability-counts';
@import 'soh-map';
