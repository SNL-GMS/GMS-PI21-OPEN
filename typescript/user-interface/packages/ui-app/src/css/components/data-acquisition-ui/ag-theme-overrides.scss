.ag-theme-dark--soh {
  border-radius: $cell-cap-radius;

  .ag-header {
    background: transparent;
  }

  .ag-header-cell {
    .ag-react-container {
      width: 100%;
    }
  }
  .ag-header-viewport {
    background-color: var(--gms-prominent-background, $gms-prominent-background);
  }

  .ag-row {
    overflow: hidden;
    border-radius: $cell-cap-radius;
  }

  .ag-ltr .ag-cell:not(.ag-cell-focus).ag-cell-last-left-pinned,
  .ag-ltr .ag-row.ag-cell-last-left-pinned,
  .ag-horizontal-left-spacer {
    border-right: none;
  }

  &:not(.table--cell-selection-only) {
    .ag-row-hover .ag-cell {
      @include brighten;
    }
  }
  &.table--cell-selection-only {
    .ag-cell:hover {
      @include brighten;
    }
  }

  // TODO: Make this dynamic so it doesn't expect a fixed
  // number of columns
  .ag-row-selected {
    .ag-cell .table-cell {
      box-sizing: border-box;
      @include selectedBorder('top', 'bottom');
    }
    .ag-cell[aria-colindex='1'] .table-cell {
      @include selectedBorder('left');
    }
    .ag-cell .table-cell[data-col-position='last'] {
      @include selectedBorder('right');
    }
  }

  .ag-body-viewport,
  .ag-root-wrapper {
    background: transparent;
  }
}
