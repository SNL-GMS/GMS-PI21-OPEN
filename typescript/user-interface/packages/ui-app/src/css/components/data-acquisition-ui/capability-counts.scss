.stacked-counts {
  @include flexColumn;
  @include centerChildren;
  &__entry {
    padding: 0;
    margin: 0;
    line-height: 1em;
    padding-right: 0.25rem;
    margin-left: -0.25rem;
    font-weight: 400;
    opacity: 0.8;
    font-size: 0.875em;
    & + & {
      margin-top: 0.125rem;
    }
  }
}

// Color the marginal capability count yellow if it is inside of a red cell
[data-cell-status='bad'] .stacked-counts__entry[data-capability-status='marginal'] {
  color: $marginal-color;
}
