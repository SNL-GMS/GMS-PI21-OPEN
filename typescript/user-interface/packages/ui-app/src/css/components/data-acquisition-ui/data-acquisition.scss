.form-field-small {
  width: 60%;
}

.form-field-large {
  margin-left: 5%;
  width: 80%;
}

.footer-button-row {
  position: absolute;
  top: 90%;
  left: 1.5%;
}

.footer-button {
  position: relative;
  display: inline;
}

.footer-button-middle {
  position: relative;
  display: inline;
  left: 200%;
}

.footer-button-right {
  position: relative;
  display: inline;
  left: 225%;
}

.station-configuration-form__row {
  display: flex;
  flex-direction: row;
  margin-bottom: 1%;
}

.station-configuration-form {
  display: flex;
  position: absolute;
  top: 0%;
  height: 20%;
  right: 0%;
  left: 0.5%;
  flex-direction: column;
  width: 100%;
  overflow-y: scroll;
}

.station-configuration-inner-menu {
  position: absolute;
  top: 45%;
  height: 20%;
  right: 0%;
  left: 1.5%;
}

.station-configuration-bottom-menu {
  position: absolute;
  top: 68%;
  height: 20%;
  right: 0%;
  left: 1.5%;
}

.station-configuration-form__labels-top-left {
  margin-left: 1% !important;
  margin-top: 0.5% !important;
  margin-right: 1% !important;
  font-size: small;
}

.station-configuration-form__labels-top-right {
  margin-left: 1% !important;
  margin-top: 1% !important;
  font-size: small;
}

.station-configuration-form__labels-middle {
  margin-left: 1% !important;
  margin-right: 15px !important;
  font-size: small;
}

.station-configuration-form__field-top {
  margin-left: 1.5% !important;
  margin-top: 15px !important;
  font-size: small;
}

.station-configuration-form__field-middle {
  margin-left: 12px !important;
  margin-top: 5px !important;
  font-size: small;
  width: 200%;
}

.station-configuration-form__field-top {
  margin-left: 5px !important;
  margin-top: 5px !important;
  font-size: small;
}

.station-configuration-form__field-bottom {
  margin-left: 3px !important;
  margin-top: 5px !important;
  font-size: small;
}

.station-configuration-form__labels-bottom-right {
  margin-left: 1% !important;
  font-size: small;
}

.station-configuration-form__labels-bottom-left {
  margin-left: 1% !important;
  font-size: small;
}

.station-configuration-form__input-middle {
  padding-left: 10% !important;
  width: 200%;
}

.configure-station-groups-select-network {
  position: relative;
  top: 0%;
  height: 100%;
  width: 40%;
  right: 0%;
  left: 0%;
}

.configure-station-groups-select-network-table {
  position: relative;
  top: 0%;
  height: 100%;
  max-height: calc(100% - 49px);
  width: 100%;
  right: 0%;
  left: 0%;
}

.configure-station-groups-edit-stations {
  position: absolute;
  top: 0%;
  height: 98%;
  right: 0%;
  left: 41%;
  visibility: hidden;
}

.configure-station-groups-no-network-selected {
  position: absolute;
  top: 0%;
  height: 98%;
  right: 0%;
  left: 41%;
  visibility: visible;
}

.configure-station-groups-associated-stations {
  position: absolute;
  top: 0%;
  height: 100%;
  width: 45%;
  right: 0%;
  left: 0%;
}

.configure-station-groups-arrows {
  position: absolute;
  top: 40%;
  height: 100%;
  width: 4%;
  right: 0%;
  left: 48%;
}

.configure-station-groups-available-stations {
  position: absolute;
  top: 0%;
  height: 100%;
  width: 45%;
  right: 0%;
  left: 55%;
}

.data-acquisition-popover {
  font-size: small !important;
  max-height: 300px !important;
  overflow-y: auto !important;
  margin: 0.5rem;
  display: inline-table;
}

.data-acquisition-import-css-popover-with-table {
  font-size: small !important;
  max-height: 500px !important;
  overflow-y: scroll !important;
  margin: 0.5rem;
}

.data-acquisition-inputs input {
  border: 1px solid var(--gms-main, $gms-main);
}

.data-acquisition-inputs input:focus {
  outline: none !important;
  border: 1px solid $data-acq-strong-warning;
}

.data-acquisition-cell .ag-cell {
  display: flex;
  justify-content: center;
  align-items: center !important;
}

.soh-update-time-controls {
  flex: 0 0 auto;
  width: 100%;
  display: flex;
  justify-content: space-around;
  align-items: center;
  color: var(--gms-main, $gms-main) !important;
}

.soh-update-time-controls__label {
  margin-bottom: 0rem !important;
  margin-left: 1rem !important;
  color: var(--gms-main, $gms-main);
}

.soh-update-time {
  font-family: $gms-mono;
  color: var(--gms-main, $gms-main);
}

.text-opacity-80 {
  opacity: 0.8;
}

.soh-drill-down-station-label {
  font-size: 1.5em;
  margin-top: 0.25em;
  padding-left: 0.125em;
  text-rendering: geometricPrecision;
}

.display-title {
  padding: 0.25em 0 0 0.25em;
  display: flex;
  align-items: baseline;
  &__subtitle {
    margin-left: 0.5em;
    color: var(--gms-prominent, $gms-prominent);
    font-size: 1.1rem;
    font-weight: 100;
    @include ellipsis;
  }
}

.soh-table-label {
  padding: 0.5em 0 0.25em 0;
  color: var(--gms-prominent, $gms-prominent);
  margin-left: 0.25em;
  font-size: 1.1rem;
  font-weight: 400;
  font-family: $gms-sans;
  -webkit-font-smoothing: antialiased;
}

.drop-zone__wrapper .drop-zone {
  &--highlighted {
    background-color: var(--gms-table-selection, $gms-table-selection);
  }
}
