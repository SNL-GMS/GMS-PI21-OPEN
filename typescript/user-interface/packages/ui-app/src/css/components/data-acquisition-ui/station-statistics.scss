/* 
  This is a workaround causing a wrapper to handle the scroll, rather than the 
  table itself
*/
.station-statistics-table__wrapper {
  border-radius: $rounded-rectangle-radius;

  overflow: auto;
  flex-grow: 999;
  background-color: var(--gms-prominent-background, $gms-prominent-background);
}

.station-statistics-table-container {
  position: relative;
  &--needs-attention,
  &--acknowledged {
    .ag-header-cell-label {
      padding-left: 7px;
    }
    .ag-cell-focus:focus {
      border: none !important;
    }
  }

  &--needs-attention {
    @include flexColumn;
    user-select: none;
  }

  &--acknowledged {
    @include flexColumn;
  }

  &--highlighted .station-statistics-table__wrapper {
    background-color: var(--gms-input-highlight, $gms-input-highlight);
    border-radius: $rounded-rectangle-radius + 2;
  }
  & .table-cell {
    cursor: pointer;
    .soh-cell__right-container {
      border-radius: 0;
      margin-right: -10px; // Compensate for the 10px of padding that the cell adds so this is right-aligned
    }
  }
}

.soh-divider-container {
  @include flexColumn;
  .top-component,
  .bottom-component {
    overflow: auto;
  }
  .horizontal-divider {
    margin: 0;
    width: 100%;
  }
}

.station-statistics__needs-attention-label {
  padding: 0.5em 0 0.25em 0;
  .station-statistics-table-container--acknowledged & {
    padding-top: 0;
  }
  color: var(--gms-prominent, $gms-prominent);
  margin-left: 0.25em;
  font-size: 1.1rem;
  font-weight: 100;
}

.station-statistics-table-container .table-cell {
  cursor: pointer;
}

.soh-table__drop-zone--highlight {
  & * {
    overflow: visible; // don't cut off shadows
  }
  .table-cell {
    @include addBoxShadow('md');
    overflow: visible;
  }
}

.drop-zone {
  height: 100%;
  flex-grow: 2;
  .station-statistics-table-container & {
    display: flex;
  }
  &--highlighted {
    .station-statistics-table__wrapper {
      background-color: var(--gms-input-highlight, $gms-input-highlight);
      border-radius: $rounded-rectangle-radius + 2;
    }
  }
}

.table-background-icon {
  fill: var(--gms-prominent, $gms-prominent);
  color: var(--gms-prominent, $gms-prominent);
}
