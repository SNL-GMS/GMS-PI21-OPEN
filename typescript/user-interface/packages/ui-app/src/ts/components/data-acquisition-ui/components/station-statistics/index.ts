export { StationStatisticsContainer as StationStatistics } from './station-statistics-container';
