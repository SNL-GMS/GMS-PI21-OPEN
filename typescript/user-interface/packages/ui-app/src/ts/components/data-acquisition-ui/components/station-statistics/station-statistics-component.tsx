/* eslint-disable react/destructuring-assignment */
import type { SohTypes } from '@gms/common-model';
import { Displays } from '@gms/common-model';
import { MILLISECONDS_IN_SECOND, millisToStringWithMaxPrecision } from '@gms/common-util';
import type GoldenLayout from '@gms/golden-layout';
import type { AppState, FilterableSOHTypes } from '@gms/ui-state';
import { dataAcquisitionActions, setSelectedStationIds, ssamControlApiSlice } from '@gms/ui-state';
import { UILogger } from '@gms/ui-util';
import isEqual from 'lodash/isEqual';
import memoizeOne from 'memoize-one';
import React from 'react';
import type { ConnectedProps } from 'react-redux';
import { connect } from 'react-redux';

import { BaseDisplay } from '~components/common-ui/components/base-display';
import type { WithAcknowledgeProps } from '~components/data-acquisition-ui/shared/acknowledge';

import type { StationStatisticsContextData } from './station-statistics-context';
import { StationStatisticsContext } from './station-statistics-context';
import { StationStatisticsPanel } from './station-statistics-panel';

const logger = UILogger.create(
  'GMS_LOG_SOH_STATION_STATISTICS',
  process.env.GMS_LOG_SOH_STATION_STATISTICS
);

/**
 * Mapping redux state to the properties of the component
 *
 * @param state App state, root level redux store
 */
const mapStateToProps = (state: AppState) => ({
  selectedStationIds: state.app.common.selectedStationIds,
  sohStatus: state.app.dataAcquisition.data.sohStatus,
  sohConfigurationQuery: ssamControlApiSlice.endpoints.getSohConfiguration.select()(state),
  groupSelected: state.app.dataAcquisition.stationStatisticsGroup,
  statusesToDisplay:
    state.app.dataAcquisition.filtersToDisplay[Displays.SohDisplays.STATION_STATISTICS]
});

/**
 * Mapping methods (actions and operations) to dispatch one or more updates to the redux store
 */
const mapDispatchToProps = {
  getSohConfiguration: ssamControlApiSlice.endpoints.getSohConfiguration.initiate,
  setSelectedStationIds,
  setStationStatisticsGroup: (group: string) => dispatch => {
    dispatch(dataAcquisitionActions.setStationStatisticsGroup(group));
  },
  setStatusesToDisplay: (filters: Record<FilterableSOHTypes, boolean>) => dispatch => {
    dispatch(
      dataAcquisitionActions.setFiltersToDisplay({
        list: Displays.SohDisplays.STATION_STATISTICS,
        filters
      })
    );
  }
};

export const stationStatisticsConnector = connect(mapStateToProps, mapDispatchToProps);

type ConnectedReduxProps = ConnectedProps<typeof stationStatisticsConnector>;

/**
 * Station Statistics component props
 */
export type StationStatisticsProps = ConnectedReduxProps &
  WithAcknowledgeProps & {
    glContainer?: GoldenLayout.Container;
  };

/**
 * Connected component that provides context and props to SOH Panel
 */
export class StationStatisticsComponent extends React.Component<StationStatisticsProps> {
  /*
   * Series of member variables to report lag statistics every 30 mins
   */

  /** Represents a half hour (30 minutes) in milliseconds */
  private readonly MS_HALF_HOUR: number = 1800000;

  /** the soh soh lag report period; how frequently the to report */
  private readonly sohLagReportPeriod: number = this.MS_HALF_HOUR;

  /** the last time the soh report was sent (reported) */
  private lastSohReport: number = Date.now();

  /** the min soh lag value */
  private minSohLag: number;

  /** the max soh lag value */
  private maxSohLag: number;

  /** the average soh lag value */
  private avgSohLag = 0;

  /** the soh station count - used to calculate the average */
  private sohStationsCount = 0;

  // Map to not check StationSOH already sent from API Gateway
  private readonly sohCheckDisplayCriteriaMap: Map<string, number>;

  private unsubscribe: null | (() => void) = null;

  private memoizedStationStatisticsContextData: () => StationStatisticsContextData;

  /**
   * constructor
   */
  public constructor(props: StationStatisticsProps) {
    super(props);
    this.sohCheckDisplayCriteriaMap = new Map<string, number>();
  }

  // ***************************************
  // BEGIN REACT COMPONENT LIFECYCLE METHODS
  // ***************************************

  public componentDidMount(): void {
    const { getSohConfiguration } = this.props;
    // eslint-disable-next-line @typescript-eslint/unbound-method
    const { unsubscribe } = getSohConfiguration();
    this.unsubscribe = unsubscribe;
  }

  /**
   * React lifecycle `componentDidUpdate`.
   * Called immediately after updating occurs. Not called for the initial render.
   *
   * @param prevProps the previous props
   * @param prevState the previous state
   */
  public componentDidUpdate(prevProps: StationStatisticsProps): void {
    if (!isEqual(prevProps.sohStatus, this.props)) {
      // Check SOH station entries to see if the lag exceeds the time limit
      this.checkSOHDisplayCriteria(this.props.sohStatus.stationAndStationGroupSoh.stationSoh);
    }
  }

  public componentWillUnmount(): void {
    this.unsubscribe?.();
  }

  /**
   * React lifecycle `render`.
   * Renders the component.
   */
  // eslint-disable-next-line react/sort-comp
  public render(): JSX.Element {
    this.memoizedStationStatisticsContextData = memoizeOne(() => {
      return {
        acknowledgeSohStatus: this.props.acknowledgeStationsByName,
        quietTimerMs: this.props.sohConfigurationQuery.data.acknowledgementQuietMs,
        sohStationStaleTimeMS: this.props.sohConfigurationQuery.data.sohStationStaleMs,
        updateIntervalSecs: this.props.sohConfigurationQuery.data.reprocessingPeriodSecs,
        selectedStationIds: this.props.selectedStationIds,
        setSelectedStationIds: this.props.setSelectedStationIds
      };
    }, isEqual);
    return (
      <StationStatisticsContext.Provider value={this.memoizedStationStatisticsContextData()}>
        <BaseDisplay
          glContainer={this.props.glContainer}
          className="soh-divider-container station-statistics-display drop-zone__wrapper"
          data-cy="station-statistics-display"
        >
          <StationStatisticsPanel
            statusesToDisplay={this.props.statusesToDisplay}
            setStatusesToDisplay={this.props.setStatusesToDisplay}
            stationGroups={this.props.sohStatus.stationAndStationGroupSoh.stationGroups}
            stationSohs={this.props.sohStatus.stationAndStationGroupSoh.stationSoh}
            groupSelected={this.props.groupSelected}
            setGroupSelected={this.props.setStationStatisticsGroup}
          />
        </BaseDisplay>
      </StationStatisticsContext.Provider>
    );
  }

  // ******************************************
  // END REACT COMPONENT LIFECYCLE METHODS
  // ******************************************

  /**
   * Check if SOH exceeds 1 minute display to UI criteria
   */
  private readonly checkSOHDisplayCriteria = (stationStatus: SohTypes.UiStationSoh[]): void => {
    // If UI Configuration has not loaded yet bail
    if (!this.props.sohConfigurationQuery || !this.props.sohConfigurationQuery.data) {
      return;
    }

    stationStatus.forEach(soh => {
      if (!this.sohCheckDisplayCriteriaMap.has(soh.stationName)) {
        // Set the entry, the first time we don't check
        this.sohCheckDisplayCriteriaMap.set(soh.stationName, soh.time);
      } else if (this.sohCheckDisplayCriteriaMap.get(soh.stationName) !== soh.time) {
        // Set the entry in the map so we don't check the entry again
        this.sohCheckDisplayCriteriaMap.set(soh.stationName, soh.time);

        // Get time and scale to milliseconds
        const deltaSecs = Date.now() / MILLISECONDS_IN_SECOND - soh.time;

        // add to avg
        this.avgSohLag += deltaSecs;
        // eslint-disable-next-line no-plusplus
        this.sohStationsCount++;

        // Check min and max
        if (!this.minSohLag || deltaSecs < this.minSohLag) {
          this.minSohLag = deltaSecs;
        }

        if (!this.maxSohLag || deltaSecs > this.maxSohLag) {
          this.maxSohLag = deltaSecs;
        }
      }
    });

    // Check if need to send statistic report
    if (Date.now() > this.lastSohReport + this.sohLagReportPeriod) {
      logger.info(
        `SOH lag statistics for last ${millisToStringWithMaxPrecision(this.sohLagReportPeriod)}. ` +
          `Max lag: ${millisToStringWithMaxPrecision(this.maxSohLag * MILLISECONDS_IN_SECOND)}, ` +
          `Min lag ${millisToStringWithMaxPrecision(this.minSohLag * MILLISECONDS_IN_SECOND)}, ` +
          `Avg lag ${millisToStringWithMaxPrecision(
            (this.avgSohLag / this.sohStationsCount) * MILLISECONDS_IN_SECOND
          )}`
      );

      // Reset statistics
      this.lastSohReport = Date.now();
      this.sohStationsCount = 0;
      this.avgSohLag = 0;
      this.minSohLag = undefined;
      this.maxSohLag = undefined;
    }
  };
}
