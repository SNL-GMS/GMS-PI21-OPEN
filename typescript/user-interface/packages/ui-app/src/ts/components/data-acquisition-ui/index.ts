import * as DataAcquisitionComponents from './components';

export { DataAcquisitionComponents };
