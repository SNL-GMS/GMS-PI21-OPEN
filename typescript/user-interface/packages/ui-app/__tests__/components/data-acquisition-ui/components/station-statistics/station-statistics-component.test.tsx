import { SohTypes } from '@gms/common-model';
import type { UiStationSoh } from '@gms/common-model/lib/soh/types';
import type { Container } from '@gms/golden-layout';
import { getStore } from '@gms/ui-state';
import { sohConfiguration } from '@gms/ui-state/__tests__/__data__/soh-configuration-query-data';
import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import type { StationStatisticsProps } from '../../../../../src/ts/components/data-acquisition-ui/components/station-statistics/station-statistics-component';
import { StationStatisticsComponent } from '../../../../../src/ts/components/data-acquisition-ui/components/station-statistics/station-statistics-component';
import { testStationSoh } from '../../../../__data__/data-acquisition-ui/soh-overview-data';
import { useQueryStateResult } from '../../../../__data__/test-util-data';

const glWidth = 1000;
const glHeight = 500;

// messing with normal stuff
const gLContainer: Container = {
  // Container
  width: glWidth,
  height: glHeight,
  parent: undefined,
  tab: undefined,
  title: 'container-title',
  layoutManager: undefined,
  isHidden: false,
  setState: jest.fn(),
  extendState: jest.fn(),
  getState: jest.fn(),
  getElement: jest.fn(),
  hide: jest.fn(),
  show: jest.fn(),
  setSize: jest.fn(),
  setTitle: jest.fn(),
  close: jest.fn(),
  // EventEmitter
  on: jest.fn(),
  emit: jest.fn(),
  trigger: jest.fn(),
  unbind: jest.fn(),
  off: jest.fn()
};

const setSelectedStationIds = jest.fn();

const selectedStationIds = ['A'];

const reduxProps: any = {
  glContainer: gLContainer,
  selectedStationIds,
  setSelectedStationIds
};

const sohStationAndGroupStatusQuery: any = {
  loading: false,
  stationAndStationGroupSoh: {
    stationSoh: [
      {
        stationName: 'A'
      }
    ]
  }
};

const sohConfigurationQuery = useQueryStateResult;
sohConfigurationQuery.data = sohConfiguration;

const channel: SohTypes.ChannelSoh = {
  allSohMonitorValueAndStatuses: [
    {
      status: SohTypes.SohStatusSummary.GOOD,
      value: 1,
      valuePresent: true,
      monitorType: SohTypes.SohMonitorType.ENV_ZEROED_DATA,
      hasUnacknowledgedChanges: false,
      contributing: false,
      thresholdMarginal: 1,
      thresholdBad: 10,
      quietUntilMs: 1
    }
  ],
  channelName: 'AAA111',
  channelSohStatus: SohTypes.SohStatusSummary.GOOD
};

const sohStatus: any = {
  loading: false,
  stationAndStationGroupSoh: {
    stationSoh: [
      {
        channelSohs: [channel],
        stationName: 'A'
      }
    ]
  }
};

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();
describe('Station Statistics Component', () => {
  const props: StationStatisticsProps = {
    ...reduxProps,
    sohStatus,
    sohConfigurationQuery,
    sohStationAndGroupStatusQuery,
    saveStationGroupSohStatus: jest.fn(),
    acknowledgeStationsByName: jest.fn(async () => new Promise(jest.fn())),
    mutate: undefined,
    result: undefined
  };

  const propsWithNoConfigQuery: StationStatisticsProps = {
    ...reduxProps,
    sohStatus,
    sohStationAndGroupStatusQuery,
    saveStationGroupSohStatus: jest.fn(),
    acknowledgeStationsByName: jest.fn(async () => new Promise(jest.fn())),
    mutate: undefined,
    result: undefined
  };

  const stationStatistics: any = new StationStatisticsComponent(props);
  const stationStatisticsNoConfigQuery: any = new StationStatisticsComponent(
    propsWithNoConfigQuery
  );
  const stationSoh: UiStationSoh = testStationSoh;
  it('is defined', () => {
    expect(StationStatisticsComponent).toBeDefined();
  });

  it.skip('can shallow render', () => {
    const { container } = render(
      <Provider store={getStore()}>
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <StationStatisticsComponent {...props} />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('Check SOH Display Criteria can handle when config query is not ready', () => {
    stationStatisticsNoConfigQuery.checkSOHDisplayCriteria([stationSoh]);
    expect(stationStatisticsNoConfigQuery.sohCheckDisplayCriteriaMap.size).toEqual(0);
  });

  it('can Check SOH Display Criteria with no entries', () => {
    stationStatistics.checkSOHDisplayCriteria([stationSoh]);
    expect(stationStatistics.sohCheckDisplayCriteriaMap.size).toEqual(1);
  });

  it('can Check SOH Display Criteria with an entry', () => {
    stationSoh.time = 1234567;
    stationStatistics.lastSohReport = 0;
    stationStatistics.sohLagReportPeriod = 0;
    stationStatistics.checkSOHDisplayCriteria([stationSoh]);
    expect(stationStatistics.sohCheckDisplayCriteriaMap.size).toEqual(1);
  });

  it('can run componentDidUpdate which adds to the map', () => {
    stationStatistics.componentDidUpdate(props);
    expect(stationStatistics.sohCheckDisplayCriteriaMap.size).toEqual(2);
  });
});
