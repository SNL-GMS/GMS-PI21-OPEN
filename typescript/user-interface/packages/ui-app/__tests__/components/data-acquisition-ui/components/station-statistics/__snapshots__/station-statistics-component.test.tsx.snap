// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Station Statistics Component can shallow render 1`] = `
<ContextProvider
  value={
    Object {
      "getServerState": undefined,
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <StationStatisticsComponent
    acknowledgeStationsByName={[MockFunction]}
    glContainer={
      Object {
        "close": [MockFunction],
        "emit": [MockFunction],
        "extendState": [MockFunction],
        "getElement": [MockFunction],
        "getState": [MockFunction],
        "height": 500,
        "hide": [MockFunction],
        "isHidden": false,
        "layoutManager": undefined,
        "off": [MockFunction],
        "on": [MockFunction],
        "parent": undefined,
        "setSize": [MockFunction],
        "setState": [MockFunction],
        "setTitle": [MockFunction],
        "show": [MockFunction],
        "tab": undefined,
        "title": "container-title",
        "trigger": [MockFunction],
        "unbind": [MockFunction],
        "width": 1000,
      }
    }
    saveStationGroupSohStatus={[MockFunction]}
    selectedStationIds={
      Array [
        "A",
      ]
    }
    setSelectedStationIds={[MockFunction]}
    sohConfigurationQuery={
      Object {
        "currentData": undefined,
        "data": Object {
          "acknowledgementQuietMs": 300000,
          "availableQuietTimesMs": Array [
            300000,
            900000,
            3600000,
            86400000,
            604800000,
          ],
          "displayedStationGroups": Array [
            "ALL_1",
            "ALL_2",
            "A_TO_H",
            "I_TO_Z",
            "EurAsia",
            "OthCont",
            "IMS_Sta",
            "CD1.1",
            "CD1.0",
            "MiniSD",
            "GSE",
            "Primary",
            "Second",
            "AuxFast",
            "AuxDel",
            "SEISMIC",
            "INFRA",
            "HYDRO",
          ],
          "historicalSamplesPerChannel": 50000,
          "maxHistoricalQueryIntervalSizeMs": 432000000,
          "redisplayPeriodMs": 5000,
          "reprocessingPeriodSecs": 20,
          "rollupStationSohTimeToleranceMs": 600000,
          "sohHistoricalTimesMs": Array [
            21600000,
            43200000,
            86400000,
          ],
          "sohStationStaleMs": 300000,
        },
        "endpointName": undefined,
        "error": undefined,
        "fulfilledTimeStamp": undefined,
        "isError": false,
        "isFetching": false,
        "isLoading": false,
        "isSuccess": true,
        "isUninitialized": true,
        "originalArgs": undefined,
        "requestId": undefined,
        "startedTimeStamp": undefined,
        "status": undefined,
      }
    }
    sohStationAndGroupStatusQuery={
      Object {
        "loading": false,
        "stationAndStationGroupSoh": Object {
          "stationSoh": Array [
            Object {
              "stationName": "A",
            },
          ],
        },
      }
    }
    sohStatus={
      Object {
        "loading": false,
        "stationAndStationGroupSoh": Object {
          "stationSoh": Array [
            Object {
              "channelSohs": Array [
                Object {
                  "allSohMonitorValueAndStatuses": Array [
                    Object {
                      "contributing": false,
                      "hasUnacknowledgedChanges": false,
                      "monitorType": "ENV_ZEROED_DATA",
                      "quietUntilMs": 1,
                      "status": "GOOD",
                      "thresholdBad": 10,
                      "thresholdMarginal": 1,
                      "value": 1,
                      "valuePresent": true,
                    },
                  ],
                  "channelName": "AAA111",
                  "channelSohStatus": "GOOD",
                },
              ],
              "stationName": "A",
            },
          ],
        },
      }
    }
  />
</ContextProvider>
`;
